README.md
# lukemadera:social-share

Cross platform (inc. Cordova) social sharing for facebook, twitter, pinterest


## Demo

[Demo](http://lukemadera-packages.meteor.com/social-share-basic)

[Source](https://github.com/lukemadera/meteor-packages/tree/master/social-share/basic)


## Dependencies

[none]


## Installation

In a Meteor app directory:
```bash
meteor add lukemadera:social-share
```


## Usage

```html
{{> lmSocialShare opts=opts}}
```

```js
if(Meteor.isClient) {
  Template.socialShareBasic.helpers({
    opts: function() {
      var opts ={
        facebook: true,
        twitter: true,
        pinterest: false,
        shareData: {
          url: 'http://google.com'
        },
        buttonHtml: {
          twitter: 'Twitter'
        }
      };
      return opts;
    }
  });
}
```

Then create your meta tags as needed. An example Iron.router meta tag script is in `router-social.js`.
You can test your meta tags with the Chrome Inspector by spoofing the User Agent to `Facebot` or `Twitterbot`.

For Cordova, add access rules to allow the links:

```js
// Social sharing
App.accessRule('*://*.facebook.com/*');
App.accessRule('*://*.fbcdn.net/*');
App.accessRule('*://*.gmail.com/*');
App.accessRule('*://*.google.com/*');
App.accessRule('*://*.linkedin.com/*');
App.accessRule('*://*.pinterest.com/*');
App.accessRule('*://*.twitter.com/*');

App.accessRule('mailto:*');
App.accessRule('sms:*');
```


CHANGELOG.md
# 1.2.0

## Features

- add `sms` and `googlePlus` type
- test and handle issues on platforms (web, Android web, Android Cordova, iOS web, iOS Cordova)
  - facebook iOS Cordova freezes so disabled.
  - sms & email iOS Cordova direct links bug so used window popup.
  - facebookMessage not supported on mobile so disabled.
  - gmail does not work well on mobile so use email instead.
- update README for meta tags example and Cordova access rules.

## Bug Fixes

- Get Cordova app links working.


# 1.1.1 (2015-12-29)

## Features

- add `url` new type
- add support for `body` shareData (for email, gmail pre-populating of body)
- remove React from Readme as no longer supported


# 1.1.0 (2015-12-28)

## Features

- add in 4 new types: facebookMessage, gmail, email, linkedIn
- add in default svg's and styling
- refactor to support custom HTML for buttons
- refactor to prepare for url shortening on click (asynchronous)

## Breaking Changes

- remove React support (did not maintain it with the updates above and do not want to bloat this package with extra dependencies).

package.js
Package.describe({
  name: 'lukemadera:social-share',
  version: '1.2.0',
  // Brief, one-line summary of the package.
  summary: 'Cross platform (inc. Cordova) social sharing for facebook, twitter, pinterest',
  // URL to the Git repository containing the source code for this package.
  git: '',
  // By default, Meteor will default to using README.md for documentation.
  // To avoid submitting documentation, set this field to null.
  documentation: 'README.md'
});

Package.onUse(function(api) {
  api.versionsFrom('1.1.0.2');

  api.use('templating@1.0.0');
  api.use('blaze@2.0.0');
  api.use('reactive-var@1.0.5');
  // api.use('react@0.14.2', 'client');

  Cordova.depends({
    'org.apache.cordova.inappbrowser': '0.6.0'
  });

  api.addFiles([
    'social-share.html',
    'social-share.css',
    'social-share-vars.js',
    'social-share-svg.js',
    'social-share.js',
    // 'React-LmSocialShare.jsx'
  ], 'client');

  api.export('lmSocialShare');
  // api.export('LmSocialShare', 'client');
});

Package.onTest(function(api) {
  api.use('tinytest');
  api.use('lukemadera:social-share');
  api.addFiles('social-share-tests.js');
});


router-social.js
if(Meteor.isServer) {

/**
  Social meta tag dynamic page serving.

  Facebook:
  - http://stackoverflow.com/questions/7656151/multiple-facebook-opengraph-objects-on-the-same-page
  - https://developers.facebook.com/docs/opengraph/howtos/maximizing-distribution-media-content
  - https://developers.facebook.com/tools/debug/og/object/
  - https://developers.facebook.com/docs/plugins/checklist

  Twitter:
  - https://dev.twitter.com/docs/cards/types/photo-card
  - https://dev.twitter.com/docs/cards/getting-started
  - https://dev.twitter.com/docs/cards/validation/validator
  NOTE: apparently Twitter Cards have to have EACH domain "approved" for cards to be used, AND it says a "few weeks" time until approval! So until the domain is approved, the cards won't show up/work!

  Google Plus:
  - https://developers.google.com/+/web/snippet/
  - https://developers.google.com/structured-data/testing-tool/
  - http://webmasters.stackexchange.com/questions/25581/how-does-google-plus-select-an-image-from-a-shared-link
*/
Router.route('metatags', {
  where: 'server',
  path: /.*/,   //match all routes / any route
  action: function() {
    var req =this.request;
    var res =this.response;
    var userAgent =req.headers['user-agent'];
    var socialScraper =false;
    if(userAgent !==undefined && (userAgent.indexOf('facebookexternalhit') >-1 || userAgent.indexOf('Facebot') >-1)) {
      socialScraper ='facebook';
    }
    else if(userAgent !==undefined && (userAgent.indexOf('Twitterbot') >-1)) {
      socialScraper ='twitter';
    }
    else if(userAgent !==undefined && (userAgent.indexOf('Google') >-1 || userAgent.indexOf('Googlebot') >-1)) {
      socialScraper ='google';
    }
    else if(userAgent !==undefined && (userAgent.indexOf('LinkedInBot') >-1)) {
      socialScraper ='linkedIn';
    }
    

    if(socialScraper) {
      var reqUrl =req.url;
      if(reqUrl[0] =='/') {
        reqUrl =reqUrl.slice(1, reqUrl.length);
      }
      var reqUrlRaw =reqUrl;    //save any URL params
      //slice off any url params
      if(reqUrl.indexOf('?') >-1) {
        reqUrl =reqUrl.slice(0, reqUrl.indexOf('?'));
      }
      if(reqUrl.indexOf('&') >-1) {
        reqUrl =reqUrl.slice(0, reqUrl.indexOf('&'));
      }
      if(reqUrl[(reqUrl.length-1)] ==='/') {
        reqUrl =reqUrl.slice(0, (reqUrl.length-1));
      }
     
      //form site url base
      var cfgAppInfo =Config.appInfo({});
      var url1 =cfgAppInfo.rootUrl;

      var twitterHandle =cfgAppInfo.twitter.handle ? '@'+cfgAppInfo.twitter.handle : '';
      var facebookAppId =cfgAppInfo.facebook.appId || '';

      var meta ={
        url: url1+'/'+reqUrlRaw,
        title: 'title',
        desc: 'desc',
        image: url1+'/img/here.jpg',
        // hack to get image to show up first time:
        // http://stackoverflow.com/a/27424085/5551755
        imageWidth: 500,
        imageHeight: 500
      };

      // Get metadata from url info.

      // TODO

      var html ='';
      html ="<!DOCTYPE html>"+
      "<html lang='en'>"+
      "<head>";
        
      if(socialScraper ==='facebook' || socialScraper ==='linkedIn') {
        html +="<meta property='og:title' content='"+meta.title+"' />"+
        "<meta property='og:site_name' content='"+cfgAppInfo.name+"' />"+
        "<meta property='og:url' content='"+meta.url+"' />"+
        "<meta property='og:description' content='"+meta.desc+"' />"+
        "<meta property='og:image' content='"+meta.image+"' />"+
        "<meta property='og:image:width' content='"+meta.imageWidth+"' />"+
        "<meta property='og:image:height' content='"+meta.imageHeight+"' />"+
        "<meta property='fb:app_id' content='"+facebookAppId+"' />";
      }
      else if(socialScraper ==='google') {
        html ="<!doctype html>"+
        "<html itemscope='' itemtype='http://schema.org/WebPage' lang='en'>"+
        "<head>";
        html +="<meta itemprop='name' property='og:title' name='title' content='"+meta.title+"' />"+
        "<meta itemprop='description' property='og:description' name='description' content='"+meta.desc+"' />"+
        "<meta itemprop='image' property='og:image' name='image' content='"+meta.image+"' />";
        // Fall backs / helpers..
        html +="<link rel='image_src' href='"+meta.image+"' />";
        html +="<title>"+meta.title+"</title>";
      }
      else if(socialScraper ==='twitter') {
        // Twitter card validator sucks and has no useful debugging info
        // but it seems many descriptions do NOT work unless we include certain
        // other descriptions. So we hack it by adding our
        // description in front of the one that makes it work..
        // Otherwise the validator just says "WARN: No metatags found".
        meta.desc =meta.desc + " NEWARK - The guest list and parade of limousines with celebrities emerging from them seemed more suited to a red carpet event in Hollywood or New York than than a gritty stretch of Sussex Avenue near the former site of the James M. Baxter Terrace public housing project here.";
        html +="<meta name='twitter:card' content='summary' />"+
        "<meta name='twitter:site' content='"+twitterHandle+"' />"+
        "<meta name='twitter:creator' content='"+twitterHandle+"' />"+
        "<meta name='twitter:url' content='"+meta.url+"' />"+
        "<meta name='twitter:title' content='"+meta.title+"' />"+
        "<meta name='twitter:description' content='"+meta.desc+"' />"+
        "<meta name='twitter:image' content='"+meta.image+"' />";
      }
      html +="</head>"+
      "<body>"+
      "</body>"+
      "</html>";
      res.writeHeader(200, {"Content-Type": "text/html"});
      res.write(html);
      res.end();
    }
    else {
      this.next();
    }
  }
});

}

social-share-svg.js
/**
Inline SVGs

## Usage:
1. Optimize SVG with SVGO
  1. https://jakearchibald.github.io/svgomg/
    1. https://github.com/svg/svgo
2. Copy SVG code inline below
  1. find and replace regex `\n` with `` to remove new lines
  2. REMOVE any height & width attributes on the svg and instead use / add `width="100%"`
3. Use by calling the template with the `name` attribute matching they key in the `svgs` object, e.g. `{{>svg name='logo'}}
  1. the SVG will be responsive and resize based on the element dimensions; use CSS as normal to set width/height (since the image default width/height are NOT set, you'll have to set for inline elements. Use the svg `viewBox` attribute to see the default dimensions.)

## References:
http://css-tricks.com/using-svg/
http://stackoverflow.com/questions/19484707/how-can-i-make-an-svg-scale-with-its-parent-container
*/

lmSocialSharePrivate.svgs ={
  'email': {
    html: '<svg viewbox="0 0 512 512"><path d="M256.4,72.5L65.8,214.8v226.4H447V214.8L256.4,72.5z M203.7,174.1c5.6-10.2,13.3-17.9,23.1-23 c9.8-5.2,20.9-7.7,33.5-7.7c10.6,0,20.1,2.1,28.4,6.2c8.3,4.2,14.6,10,19,17.7c4.3,7.6,6.6,16,6.6,25c0,10.8-3.3,20.5-10,29.2 c-8.3,11-19,16.5-32,16.5c-3.5,0-6.1-0.6-7.9-1.8c-1.8-1.2-3-3-3.6-5.4c-5,4.8-10.7,7.2-17.2,7.2c-7,0-12.8-2.4-17.5-7.2 c-4.6-4.9-7-11.3-7-19.3c0-10,2.8-19,8.4-27.2c6.8-10,15.4-15,26-15c7.5,0,13.1,2.9,16.7,8.6l1.6-7.1h16.8l-9.6,45.5 c-0.6,2.9-0.9,4.7-0.9,5.6c0,1.1,0.2,1.8,0.7,2.4c0.5,0.5,1.1,0.8,1.7,0.8c2.1,0,4.7-1.2,7.9-3.7c4.3-3.2,7.8-7.5,10.5-12.9 c2.7-5.4,4-11,4-16.8c0-10.4-3.7-19.1-11.3-26.1c-7.5-7-18.1-10.5-31.5-10.5c-11.5,0-21.2,2.3-29.1,7c-7.9,4.6-13.9,11.2-18,19.7 c-4,8.4-6,17.3-6,26.5c0,8.9,2.3,17,6.8,24.3c4.5,7.3,10.8,12.7,19,16c8.1,3.3,17.5,5,27.9,5c10.1,0,18.8-1.4,26.1-4.2 c7.3-2.8,13.1-7,17.6-12.7h13.5c-4.2,8.6-10.7,15.3-19.4,20.2c-10,5.6-22.3,8.3-36.6,8.3c-14,0-26-2.3-36.1-7.1 c-10.2-4.7-17.7-11.7-22.7-20.8c-4.9-9.2-7.4-19.2-7.4-30C195.3,195.4,198,184.4,203.7,174.1z M79.2,219.4l140.7,94.3L79.2,422.6 V219.4z M89.9,429.8l166.5-128.9l166.5,128.9H89.9z M433.6,422.6L292.9,313.7l140.7-94.3V422.6z M263.3,184.5 c2.2,2.5,3.4,6.2,3.4,11c0,4.3-0.8,8.9-2.5,13.8c-1.6,5-3.5,8.7-5.8,11.3c-1.6,1.8-3.2,3.1-4.8,4c-2.1,1.2-4.3,1.8-6.8,1.8 c-3.2,0-5.9-1.3-8.1-3.8c-2.2-2.5-3.2-6.5-3.2-11.8c0-4,0.8-8.6,2.3-13.9c1.6-5.2,3.9-9.2,7-12c3.1-2.8,6.5-4.2,10.1-4.2 C258.3,180.6,261.1,181.9,263.3,184.5z"/></svg>'
  },
  'facebook': {
    html: '<svg viewbox="0 0 512 512"><path d="M211.9 197.4h-36.7v59.9h36.7V433.1h70.5V256.5h49.2l5.2-59.1h-54.4c0 0 0-22.1 0-33.7 0-13.9 2.8-19.5 16.3-19.5 10.9 0 38.2 0 38.2 0V82.9c0 0-40.2 0-48.8 0 -52.5 0-76.1 23.1-76.1 67.3C211.9 188.8 211.9 197.4 211.9 197.4z"/></svg>'
  },
  'facebookMessage': {
    html: '<svg viewbox="0 0 512 512"><path d="M255,68.8C152.8,68.8,69.9,146,69.9,241.2c0,54.3,26.9,102.7,69,134.3v65.7l63-34.8c16.8,4.7,34.6,7.2,53.1,7.2 c102.2,0,185.1-77.2,185.1-172.4S357.2,68.8,255,68.8z M273.4,301l-47.1-50.6l-92,50.6l101.2-108.1l48.3,50.6l90.8-50.6L273.4,301z"/></svg>'
  },
  'gmail': {
    html: '<svg viewbox="0 0 512 512"><path d="M451.2,365.2l-0.1-209l0,0l0,0v-9.3c0-17.4-14.1-31.7-31.4-31.9l0,0H93.2l0,0c-17.6,0-31.9,14.4-31.9,31.9 v218.2c0,17.5,14.2,31.8,31.7,31.9l0,0h0.2h11.4h314.9v-0.6C443,394.1,450.6,385.2,451.2,365.2z M404.7,125.6l-149,109.8l-148-109.8 H404.7z M104.6,186.8L255.7,292l151.1-104.1l0,0l0.1,198.5H104.6L104.6,186.8L104.6,186.8z"/></svg>'
  },
  'googlePlus': {
    html: '<svg viewbox="0 0 512 512"><path d="M306.4,245.9c1.1,6.7,2.1,13.4,2.1,22.2c0,76.5-51.3,130.9-128.6,130.9c-74.1,0-134-59.9-134-134 s59.9-134,134-134c36.2,0,66.4,13.2,89.7,35l-36.4,35c-10-9.6-27.4-20.7-53.4-20.7c-45.7,0-83.1,37.9-83.1,84.6 c0,46.7,37.3,84.6,83.1,84.6c53,0,72.9-38.1,76-57.8h-76v-45.9H306.4z M428.7,245.9v-38.3h-38.3v38.3h-38.3v38.3h38.3v38.3h38.3 v-38.3H467v-38.3H428.7z"/></svg>'
  },
  'linkedIn': {
    html: '<svg viewbox="0 0 512 512"><path d="M186.4 142.4c0 19-15.3 34.5-34.2 34.5 -18.9 0-34.2-15.4-34.2-34.5 0-19 15.3-34.5 34.2-34.5C171.1 107.9 186.4 123.4 186.4 142.4zM181.4 201.3h-57.8V388.1h57.8V201.3zM273.8 201.3h-55.4V388.1h55.4c0 0 0-69.3 0-98 0-26.3 12.1-41.9 35.2-41.9 21.3 0 31.5 15 31.5 41.9 0 26.9 0 98 0 98h57.5c0 0 0-68.2 0-118.3 0-50-28.3-74.2-68-74.2 -39.6 0-56.3 30.9-56.3 30.9v-25.2H273.8z"/></svg>'
  },
  'pinterest': {
    html: '<svg viewbox="0 0 512 512"><path d="M266.6 76.5c-100.2 0-150.7 71.8-150.7 131.7 0 36.3 13.7 68.5 43.2 80.6 4.8 2 9.2 0.1 10.6-5.3 1-3.7 3.3-13 4.3-16.9 1.4-5.3 0.9-7.1-3-11.8 -8.5-10-13.9-23-13.9-41.3 0-53.3 39.9-101 103.8-101 56.6 0 87.7 34.6 87.7 80.8 0 60.8-26.9 112.1-66.8 112.1 -22.1 0-38.6-18.2-33.3-40.6 6.3-26.7 18.6-55.5 18.6-74.8 0-17.3-9.3-31.7-28.4-31.7 -22.5 0-40.7 23.3-40.7 54.6 0 19.9 6.7 33.4 6.7 33.4s-23.1 97.8-27.1 114.9c-8.1 34.1-1.2 75.9-0.6 80.1 0.3 2.5 3.6 3.1 5 1.2 2.1-2.7 28.9-35.9 38.1-69 2.6-9.4 14.8-58 14.8-58 7.3 14 28.7 26.3 51.5 26.3 67.8 0 113.8-61.8 113.8-144.5C400.1 134.7 347.1 76.5 266.6 76.5z"/></svg>'
  },
  'sms': {
    html: '<svg viewbox="0 0 512 512"><path d="M427.2,189.5L427.2,189.5C386,95.5,276.2,52.1,182.1,93.4C87.6,134.1,44.7,243.9,85.5,338 c8.3,19.9,19.9,37.1,33.4,51.7c-3.1,18.8-21.4,38.7-53.8,59.1c34,2.1,62.2-8.9,84.7-32.4c51.2,36.1,119.2,45,180.8,18.3 C425.2,393.9,468,284.1,427.2,189.5z M179.5,286.6c-3.1,4.2-6.8,7.8-12,9.9c-5.2,2.1-11.5,3.2-17.8,3.2c-11,0-19.3-2.1-25.1-6.8 c-5.8-4.2-9.4-11-11-19.9l0,0l0,0c0,0,0,0,0,0l14.1-2.1c1.1,5.8,3.2,9.9,6.8,13.1c3.7,2.6,8.4,4.2,15.1,4.2c6.3,0,11-1,14.1-3.6 c3.7-3.1,5.2-6.3,5.2-9.4c0-3.6-1.5-5.7-4.2-7.8c-2.1-1-6.8-2.6-14.6-4.7c-10.5-2.6-17.8-4.7-22-6.8c-3.6-2.1-6.8-4.7-8.9-7.8 c-2.1-3.6-3.1-7.3-3.1-11.5c0-3.7,0.5-7.3,2.6-10.5c1.6-3.1,3.7-5.7,6.8-7.8c2.1-1.6,5.2-3.1,8.9-4.2c4.2-1,7.8-1.6,12.5-1.6 c6.3,0,12,1.1,17.2,2.6c5.2,2.1,8.9,4.7,11,7.8c2.6,3.1,4.2,7.3,4.7,13.1l-13.6,2.1c-0.5-4.7-2.6-7.9-5.8-10.5 c-2.6-2.1-7.3-3.6-12.5-3.6c-6.8,0-11.5,1.1-14.1,3.1c-2.6,2.6-4.2,4.7-4.2,7.9c0,1.6,0.5,3.6,1.6,4.7c1.6,1.5,3.2,3.1,5.8,4.2 c1.6,0.5,5.8,1.6,12.5,3.7c9.9,2.6,16.7,4.7,20.9,6.3c4.2,1.6,7.3,4.2,9.4,7.8c2.1,3.2,3.2,6.8,3.2,12 C183.2,278.3,182.1,282.4,179.5,286.6z M313.8,298.2h-14.1v-52.8c0-5.8-0.5-9.9-1.6-12.5c-0.5-2.1-2.6-4.2-4.7-5.8 c-2.6-1.6-5.2-2.6-8.4-2.6c-6.3,0-11,2.1-15.1,6.3c-3.7,3.7-5.8,10-5.8,18.8v48.6h-14.1v-54.4c0-6.3-1.1-11-3.6-14.6 c-2.1-3.2-5.8-4.7-11-4.7c-4.2,0-7.8,1.1-11.5,3.1c-3.1,2.6-5.7,5.2-7.3,9.4c-1.6,4.1-2.1,9.9-2.1,17.8v43.4h-14.1l0,0v-83.6H213 V226c2.6-4.2,6.3-7.3,10.5-9.9c4.7-2.6,9.4-3.7,15.1-3.7c6.3,0,11,1,15.2,3.7c4.2,2.6,6.8,6.3,8.3,11c6.8-9.9,15.2-14.6,26.1-14.6 c7.9,0,14.6,2.1,18.8,6.8c4.7,4.7,6.8,11.5,6.8,21.5V298.2z M394.8,286.6c-2.6,4.2-6.8,7.8-12,9.9c-5.2,2.1-11,3.2-17.8,3.2 c-11,0-19.3-2.1-25.1-6.8c-5.2-4.2-8.9-11-10.4-19.9l0,0l0,0c0,0,0,0,0,0l13.6-2.1c1,5.8,3.1,9.9,6.8,13.1c3.6,2.6,8.9,4.2,15.2,4.2 c6.3,0,11.5-1,14.6-3.6c3.2-3.1,4.7-6.3,4.7-9.4c0-3.6-1.6-5.7-4.2-7.8c-2.1-1-6.8-2.6-14.6-4.7c-10.4-2.6-17.8-4.7-21.4-6.8 c-4.2-2.1-7.3-4.7-9.4-7.8c-2.1-3.6-3.1-7.3-3.1-11.5c0-3.7,1.1-7.3,2.6-10.5c1.6-3.1,4.2-5.7,6.8-7.8c2.6-1.6,5.2-3.1,9.4-4.2 c3.7-1,7.8-1.6,12-1.6c6.8,0,12.6,1.1,17.2,2.6c5.2,2.1,8.9,4.7,11,7.8c2.6,3.1,4.2,7.3,5.2,13.1l-14.1,2.1 c-0.5-4.7-2.6-7.9-5.2-10.5c-3.1-2.1-7.3-3.6-13-3.6c-6.3,0-11,1.1-14.1,3.1c-2.6,2.6-4.2,4.7-4.2,7.9c0,1.6,0.5,3.6,2.1,4.7 c1.1,1.5,2.6,3.1,5.2,4.2c1.6,0.5,5.7,1.6,12.5,3.7c9.9,2.6,17.2,4.7,20.9,6.3c4.2,1.6,7.3,4.2,9.4,7.8c2.6,3.2,3.7,6.8,3.7,12 C399,278.3,397.5,282.4,394.8,286.6z"/></svg>'
  },
  'twitter': {
    html: '<svg viewbox="0 0 512 512"><path d="M419.6 168.6c-11.7 5.2-24.2 8.7-37.4 10.2 13.4-8.1 23.8-20.8 28.6-36 -12.6 7.5-26.5 12.9-41.3 15.8 -11.9-12.6-28.8-20.6-47.5-20.6 -42 0-72.9 39.2-63.4 79.9 -54.1-2.7-102.1-28.6-134.2-68 -17 29.2-8.8 67.5 20.1 86.9 -10.7-0.3-20.7-3.3-29.5-8.1 -0.7 30.2 20.9 58.4 52.2 64.6 -9.2 2.5-19.2 3.1-29.4 1.1 8.3 25.9 32.3 44.7 60.8 45.2 -27.4 21.4-61.8 31-96.4 27 28.8 18.5 63 29.2 99.8 29.2 120.8 0 189.1-102.1 185-193.6C399.9 193.1 410.9 181.7 419.6 168.6z"/></svg>'
  }
};

social-share.js
/**

TODO
- add native (cordova) app support / dialogs, especially for facebook (message)

Known issues:

- facebook does not close properly on ios Cordova so we disable it.
- ios Cordova sms & email links bug so need to open separate window. Still
 works, but not as clean.
 https://forums.meteor.com/t/whitelisting-mailto-links-in-cordova/13671
- facebook message is not supported on mobile, so we disable it.
 https://developers.facebook.com/docs/sharing/reference/send-dialog
- hyphens in URLs prevent meta data (image) from showing up on Google Plus on web.
- gmail does not work well on mobile (it opens, but no pre-populated content).
 So we just use `email` instead. On web gmail will still show up.


NOTE: These will NOT work properly on localhost. 3rd party (social) services
 (for sharing) will complain about invalid URLs because
 localhost is not publicly accessible. Must be tested live / with publicly
 accessible, valid URLs.

TODO: add url shortening (just add a custom callback function as a parameter
 - to shorten, or do anything else).
This creates social sharing buttons. There's some complications for making
 shortened url's (currently we're using bitly) in that we want to minimize
 shorten calls by only forming the link AFTER the user has clicked a particular
 button. Otherwise on page load we'd form EVERY short link, which with the
 current 7 buttons means 7 API calls, even if they are not clicked on!
 There is a browser pop-up blocker issue that happens though if an AJAX call
 is made in-between the clicking of a button and displaying a pop-up window
 so we had to work around that too.
*/

/**
@param {String} btnId The html element to attach click (and touch) event
 handlers to.
@param {String} type One of 'email', 'facebook', 'facebookMessage', 'gmail',
 'googlePlus', 'linkedIn', 'pinterest', 'sms', 'twitter', 'url'
 Note: `facebookMessage` does not work on mobile.
 Note: `email` will be used instead of gmail on mobile platforms.
 Note: `sms` is mobile only.
@param {Object} shareData
  @param {String} url
  @param {String} description [pinterest only]
  @param {String} image [pinterest only]
  @param {String} subject [email, gmail, linkedIn only]
  @param {String} body [email, gmail, linkedIn, sms only] Url will be appended to
   the end.
  @param {String} facebookAppId [facebookMessage only]
  @param {String} redirectUrl [facebookMessage only]
  @param {String} [defaultShareText] The default tweet/fb post/etc. text to
   pre-populate (user can alter).
  @param {Object} [buttonHtml] The button html, one key per each type. If not set,
   will use defaults.
    @param {String} [email]
    @param {String} [facebook]
    @param {String} [facebookMessage]
    @param {String} [gmail]
    @param {String} [googlePlus]
    @param {String} [linkedIn]
    @param {String} [pinterest]
    @param {String} [sms]
    @param {String} [twitter]
    @param {String} [url]
*/
lmSocialShare.add =function(btnId, type, shareData, params) {
  var platform =lmSocialSharePrivate.getPlatform();
  var directLink =lmSocialSharePrivate.useDirectLink(type, platform);
  if(document.getElementById(btnId)) {
    if(directLink) {
      lmSocialShare.formLink(type, shareData, function(err, result) {
        if(result.link) {
          document.getElementById(btnId).href =result.link;
          document.getElementById(btnId).target ='_blank';
        }
        else {
          console.error('lm-social-share no link');
        }
      });
    }
    else {
      document.getElementById(btnId).onclick =function(evt) {
        lmSocialSharePrivate.triggerShare(type, shareData, { btnId: btnId });
      };
    }
  }
};

lmSocialSharePrivate.useDirectLink =function(type, platform) {
  if(type === 'email' ) {
    return ( platform.cordova && platform.ios ) ? false :
     ( platform.mobile || platform.cordova ) ? true : false;
  }
  else if(type === 'sms' ) {
    return ( platform.cordova && platform.ios ) ? false : true;
  }
  return false;
};

lmSocialSharePrivate.getPlatform =function() {
  var platform ={
    cordova: Meteor.isCordova,
    ios: /iPhone|iPad|iPod/i.test(navigator.userAgent),
    android: /Android/i.test(navigator.userAgent),
    blackberry: /BlackBerry/i.test(navigator.userAgent),
    windows: /IEMobile/i.test(navigator.userAgent)
  };
  platform.mobile = ( platform.ios || platform.android || platform.blackberry
   || platform.windows ) ? true : false;

  return platform;
};

lmSocialShare.formLink =function(type, shareData, callback) {
  var link = null;
  var platform;
  var body = shareData.body ? ( shareData.body + "\n\n" + shareData.url)
   : shareData.url;
  if( type === 'email' ) {
    link ='mailto:?subject='+encodeURIComponent(shareData.subject) +
     '&body='+encodeURIComponent(body);
    return callback(null, { link: link });
  }
  else if( type === 'sms' ) {
    platform =lmSocialSharePrivate.getPlatform();
    // ios does not allow pre-populating body.
    // http://stackoverflow.com/questions/16165393/ios-sms-scheme-in-html-hyperlink-with-body
    link = ( platform.ios ) ? 'sms:' : ( 'sms:?body=' + encodeURIComponent(body) );
    return callback(null, { link: link });
  }
  else if( type === 'gmail' ) {
    link ='https://mail.google.com/mail/u/0/?view=cm&fs=1&su=' +
     encodeURIComponent(shareData.subject) + '&body=' +
     encodeURIComponent(body);
    return callback(null, { link: link });
  }
  else if( type === 'googlePlus' ) {
    link ='https://plus.google.com/share?url=' +
     encodeURIComponent(shareData.url);
    return callback(null, { link: link });
  }
  else if( type === 'facebook' ) {
    link ='https://www.facebook.com/sharer/sharer.php?u=' +
     encodeURIComponent(shareData.url);
    return callback(null, { link: link });
  }
  else if( type === 'facebookMessage' ) {
    link ='https://www.facebook.com/dialog/send?link=' +
     encodeURIComponent(shareData.url) + '&app_id=' + shareData.facebookAppId +
     '&redirect_uri=' + encodeURIComponent(shareData.redirectUrl)
     + '&display=popup';
    return callback(null, { link: link });
  }
  else if( type === 'linkedIn' ) {
    link ='https://www.linkedin.com/shareArticle?mini=true&url=' +
     encodeURIComponent(shareData.url) + '&title=' +
     encodeURIComponent(shareData.subject) + '&summary=' +
     encodeURIComponent(body);
    return callback(null, { link: link });
  }
  else if( type === 'pinterest' ) {
    link ='http://pinterest.com/pin/create/button/?url=' +
     encodeURIComponent(shareData.url) + '&media=' +
     encodeURIComponent(shareData.image) + '&description=' +
     encodeURIComponent(shareData.description);
    return callback(null, { link: link });
  }
  else if( type === 'twitter' ) {
    link ='https://twitter.com/share?url=' + encodeURIComponent(shareData.url);
    if(shareData.defaultShareText !==undefined) {
      link +='&text='+encodeURIComponent(shareData.defaultShareText);
    }
    return callback(null, { link: link });
  }
  callback(null, null);
};

lmSocialSharePrivate.inputSelectAll =function(classname, id) {
  var ele =document.getElementById(id);
  // get input
  ele =ele.getElementsByClassName(classname)[0];
  if(ele) {
    ele.setSelectionRange(0, ele.value.length);
  }
};

lmSocialSharePrivate.triggerShare =function(type, shareData, params) {
  if( type === 'url' ) {
    lmSocialSharePrivate.inputSelectAll('lm-social-share-url-input', params.btnId);
    return;
  }
  if(type && shareData) {
    // Must open pop-up window immediately (before AJAX call) to prevent
    // pop-up blockers from stopping it. After it is open, then set the url.
    var windowHandle;
    if(Meteor.isCordova) {
      // Can NOT set the href after the window is open if using Cordova.
      // https://wiki.apache.org/cordova/InAppBrowser
      // windowHandle =window.open(null, "_blank", "location=yes");
      lmSocialShare.formLink(type, shareData, function(err, result) {
        if(result.link) {
          windowHandle =window.open(result.link, "_blank", "location=yes");
        }
        else {
          console.error('lm-social-share no link');
        }
      });
    }
    else {
      windowHandle =window.open(null, "", "height=440,width=640,scrollbars=yes");
      lmSocialShare.formLink(type, shareData, function(err, result) {
        if(result.link) {
          windowHandle.location.href =result.link;
        }
        else {
          windowHandle.close();
        }
      });
    }
  }
};

lmSocialSharePrivate.formButtonHtml =function(types, html, shareData) {
  html = html || {};
  types.forEach(function(key) {
    html[key] = html[key] ? html[key] : ( key === 'url' ) ? (
      "<input type='text' readonly=true value='" + shareData.url +"' class='lm-social-share-url-input' />"
    ) : (
      "<div class='lm-social-share-btn-icon'>" + lmSocialSharePrivate.svgs[key].html + "</div>"
    );
  });
  return html;
};

/**
Filters out known issues
*/
lmSocialSharePrivate.filterTypes =function(opts) {
  var platform =lmSocialSharePrivate.getPlatform();

  // Facebook does not close (freezes app) in Cordova app ios..
  // TODO - fix this
  if( opts.facebook && ( platform.ios && platform.cordova ) ) {
    opts.facebook =false;
  }

  // Facebook message does not work on mobile
  if( opts.facebookMessage && ( platform.mobile || platform.cordova ) ) {
    opts.facebookMessage =false;
  }

  // Gmail does not work well on mobile so just disable it and use email instead.
  if( opts.gmail && ( platform.mobile || platform.cordova ) ) {
    opts.gmail =false;
    opts.email =true;
  }

  // SMS only works on mobile
  if( opts.sms && ( !platform.mobile && !platform.cordova ) ) {
    opts.sms =false;
  }

  return opts;
};

Template.lmSocialShare.created =function() {
  this.instid = new ReactiveVar((Math.random() + 1).toString(36).substring(7));
};

Template.lmSocialShare.helpers({
  data: function() {
    var opts =this.opts;
    var ret ={
      eles: {},
      active: {}
    };
    var instid =Template.instance().instid.get();
    var types =['email', 'facebook', 'facebookMessage', 'gmail', 'googlePlus',
     'linkedIn', 'pinterest', 'sms', 'twitter', 'url'];
    opts =lmSocialSharePrivate.filterTypes(opts);

    var shareData =opts.shareData;
    var buttonHtml =lmSocialSharePrivate.formButtonHtml(types,
     opts.buttonHtml, shareData);
    var ii, type;
    for(ii =0; ii<types.length; ii++) {
      (function(ii) {
        type =types[ii];
        if(!opts[type]) {
          ret.active[type] =false;
        }
        else {
          ret.active[type] =true;
          ret.eles[types[ii]] ={
            id: 'socialShare'+instid+types[ii],
            buttonHtml: buttonHtml[types[ii]]
          };
          setTimeout(function() {
            lmSocialShare.add(ret.eles[types[ii]].id, types[ii], shareData, {});
          }, 500);
        }
      })(ii);
    }
    return ret;
  }
});